core/src/main/clojure/xtdb/metadata.clj

```diff
@@ -311,23 +315,47 @@
     {:col-names (into #{} col-names)
      :page-idx-cache page-idx-cache}))
 
-(defn ->table-metadata ^xtdb.metadata.ITableMetadata [^IVectorReader metadata-reader, {:keys [col-names ^Map page-idx-cache]}]
-  (reify ITableMetadata
-    (metadataReader [_] metadata-reader)
-    (columnNames [_] col-names)
-    (rowIndex [_ col-name page-idx] (.get page-idx-cache [col-name page-idx]))
-    (iidBloomBitmap [_ page-idx]
-      (let [bloom-rdr (-> (.structKeyReader metadata-reader "columns")
-                          (.listElementReader)
-                          (.structKeyReader "bloom"))]
+(defrecord TableMetadata [^HashTrie trie
+                          ^RelationReader meta-rel-reader
+                          ^ArrowBuf buf
+                          ^IVectorReader metadata-rdr
+                          col-names
+                          ^Map page-idx-cache
+                          ^AtomicInteger ref-count]
+  ITableMetadata
+  (metadataReader [_] metadata-rdr)
+  (columnNames [_] col-names)
+  (rowIndex [_ col-name page-idx] (.get page-idx-cache [col-name page-idx]))
+  (iidBloomBitmap [_ page-idx]
+    (let [bloom-rdr (-> (.structKeyReader metadata-rdr "columns")
+                        (.listElementReader)
+                        (.structKeyReader "bloom"))]
+
+      (when-let [bloom-vec-idx (.get page-idx-cache ["xt$iid" page-idx])]
+        (when (.getObject bloom-rdr bloom-vec-idx)
+          (bloom/bloom->bitmap bloom-rdr bloom-vec-idx)))))
```

Comment:
@jarohen I kept the `ITableMetadata` around but things can also be grabbed directly from the record (see trie). I could either slim down the interface (just `rowIndex` and `iidBloomBitmap`) or move everything to the protocol and make the `defrecord` a `deftype`.

Reply:
aware it's not massively Clojure-y, but I don't mind this - the strongly-typed interface means the caller doesn't need to worry as much about type hinting.

that said, column-names is just a set, so maybe not as required there. :shrug: